### DISCUSSION TITLE
Respect the `MaxThreads` of an in-out function

### DISCUSSION THREAD

**Discussion by JAicewizard** [duckdb/duckdb#9191] [Ideas]
Tagging @Mytherin as recomended by Thijs
Currently an in-out function can have a gobal state with `MaxThreads` method, which is supposed to indicate the maximum amount of threads used.
This method however is never used, at least not for in-out functions.
In the trivial case of `MaxThreads==1` this can be resolved easily by changing `PhysicalTableInOutFunction::ParallelOperator` to only return true when `MaxThreads!=1`.
In the more difficult case, where `MaxThreads>1`, this does not work, as the function will still be executed in all threads. @Tishj suggested to have `ParallelOperator` return a number instead, to indicate the max amount of threads, I guess similar to how `MaxThreads` already works.

One of these being implemented would be very nice, but I do think that since the API exposes a maximum amount of threads, this should be respected. People could set a max amount of threads equal to a certain amount of open connections, and not think duckdb would use more threads (leading to a race condition, not all data being processes, or a crash).
However I also think that the level of granularity if very high, maybe it would be best to expose less control in the API, and focus on having an API that has less weird oddities and is documented. So removing the option to specify the amount of threads, and only allow specifying parallelism or not, would also be an option.

My personal preference goes out to the second option, changing the signature of `ParallelOperator`, as it feels the most perfect and complete to me.

**Comment by Tishj** [comment]
I would like to add, there are two changes proposed here:

1. Make `PhysicalTableInOutFunction::ParallelOperator` conditional on the value returned by `GlobalTableFunctionState::MaxThreads`, returning `false` if the returned value is 1.

2. Change the prototype of `PhysicalOperator::ParallelOperator` just so we can limit the amount of parallelism  

I agree with 1, I do not think the benefit of 2 is significant enough to warrant this change.

**Reply by JAicewizard** [reply]
If 1 is the chosen option, I do think it would be nice to have it as a comment or some documentation saying somewhere that `MaxThreads` is not strictly followed when it is an in-out function

**Comment by l1t1** [comment]
what does in-out function mean?

**Reply by JAicewizard** [reply]
I think it is a special kind of table function (https://github.com/duckdb/duckdb/blob/0ce25fa35e93a3e334ac9a9949bbc515d1bbd23b/src/include/duckdb/function/table_function.hpp#L238-L243). I find it confusing that the same type is used for different kind of functions but it might have a good reason, they are related after all

**Reply by Tishj** [reply]
Table In-Out function is a table function that takes a subquery as input instead of scalars.
Hence why it turns into a regular operator rather than a source, it gets fed *by* a source
